src/librustc/query/mod.rs

```diff
@@ -310,6 +310,11 @@ rustc_queries! {
         /// Maps from a trait item to the trait item "descriptor".
         query associated_item(_: DefId) -> ty::AssocItem {}
 
+        /// Collects the associated items defined on a trait or impl.
+        query associated_items(key: DefId) -> ty::AssocItemsIterator<'tcx> {
```

Comment:
Could this be just the slice? You might need to replace this:
```rust
for x in tcx.associated_items(def_id)
```
with:
```rust
for &x in tcx.associated_items(def_id)
```

But for the most part it should be fine not to even do that.